src/webots/nodes/utils/WbNodeUtilities.cpp

```diff
@@ -1785,72 +1769,69 @@ WbNodeUtilities::Answer WbNodeUtilities::isSuitableForTransform(const WbNode *co
     return UNSUITABLE;
   }
 
-  if (isRobotTypeName(srcModelName)) {
-    if (destModelName == "Group" || destModelName == "Transform" || destModelName == "Solid" || destModelName == "Charger" ||
-        destModelName == "Connector") {
-      if (!hasSolidChildren(srcNode))
-        return LOOSING_INFO;
-
-      return UNSUITABLE;
-    }
-  }
-
   if (srcModelName == "Group" || srcModelName == "Transform") {
-    if (srcModelName == "Group" && destModelName == "Transform")
+    if (destModelName == "Solid" || (srcModelName == "Group" && destModelName == "Transform"))
       return SUITABLE;
     if (srcModelName == "Transform" && destModelName == "Group")
       return LOOSING_INFO;
 
-    if (srcNode->isTopLevel()) {
-      if (destModelName == "Solid" || destModelName == "Charger" || isRobotTypeName(destModelName))
-        return SUITABLE;
+    if (srcNode->isTopLevel())
+      return (destModelName == "Charger" || isRobotTypeName(destModelName)) ? SUITABLE : UNSUITABLE;
+
+    if (isSolidDeviceTypeName(destModelName))
+      return hasARobotAncestor(srcNode) ? SUITABLE : UNSUITABLE;
 
-      return UNSUITABLE;
-    }
-    if (destModelName == "Solid" || isSolidDeviceTypeName(destModelName)) {
-      const QString &topNodeModelName = findTopNode(srcNode)->nodeModelName();
-      const QString &parentModelName = srcNode->parentNode()->nodeModelName();
-      if (isRobotTypeName(topNodeModelName))
-        return SUITABLE;
-      if (destModelName == "Solid" && isSolidTypeName(parentModelName))
-        return SUITABLE;
-
-      return UNSUITABLE;
-    }
     return UNSUITABLE;
   }
 
-  if (isSolidTypeName(srcModelName)) {
-    if (destModelName == "Group" || destModelName == "Transform") {
-      if (!hasSolidChildren(srcNode))
-        return LOOSING_INFO;
-
-      return UNSUITABLE;
+  if (destModelName == "Group" || destModelName == "Transform") {
+    if (isSolidTypeName(srcModelName)) {
+      if (!hasDeviceChildren) {
+        // cache value
+        hasDeviceChildren = new bool[1];
```

Comment:
There is a memory leak here: `hasDeviceChildren` will never be free.

Reply:
No, there is no memory leak current in the code.
The hasDeviceChildren pointer is deleted where the `WbNodeUtilities::isSuitableForTransform` is called.

Reply:
I am pretty sure there is a memory leak here.
The `hasDeviceChildren` pointer is not passed by reference, hence it is not modified outside the scope of this method, thus the memory leak. You can check the value you are deleting afterwards, it will always be NULL.

Reply:
You are right.
Instead of fixing and passing the pointer by reference so that the value can be modified, I simplified the code and passed a reference to an int.
It is not as clean as using a pointer for storing the three states (not computer, true and false) but the code is simpler and less error-prone.